### ISSUE TITLE
Nice Work!  Ideas to improve Performance

### ISSUE THREAD

**Issue by ahotrod** [pohanchi/huggingface_albert#1]
## ❓ Questions & Help

Good job, thanks for sharing your code.

My system has 2 x NVIDIA 1080Ti.  Running data parallel doesn't work for me with the current transformers, and I'd prefer to run distributed processing using 2 x 1080Ti, but transformers currently has a limitation in that it doesn't `do_eval` with distributed processing.  So with transformers I'm forced to do a 2-step script which separates distributed 2x GPU training and single GPU evaluation:

```
#!/bin/bash

MODEL_PATH=/media/dn/dssd/nlp/huggingface_albert/examples/runs/albert_base_sqd1_dist
SQUAD_DIR=/media/dn/dssd/nlp/transformers/examples/scripts/squad1.1

python -m torch.distributed.launch --nproc_per_node=2 run_squad_albert.py \
  --train_file ${SQUAD_DIR}/train-v1.1.json \
  --predict_file ${SQUAD_DIR}/dev-v1.1.json \
  --model_type albert \
  --output_dir ${MODEL_PATH} \
  --do_train \
  --overwrite_output_dir \
  --do_lower_case \
  --warmup_steps 384 \
  --save_steps 1000 \
  --logging_steps 1000 \
  --num_train_epochs 3

CUDA_VISIBLE_DEVICES=0 python run_squad_albert.py \
  --model_type albert \
  --model_name_or_path ${MODEL_PATH} \
  --do_eval \
  --do_lower_case \
  --train_file ${SQUAD_DIR}/train-v1.1.json \
  --predict_file ${SQUAD_DIR}/dev-v1.1.json \
  --max_seq_length 384 \
  --output_dir ${MODEL_PATH}
$@
```
However, your `run_squad_albert.py` doesn't include `--model_name_or_path` functionality.  Is there another way to run `--do_eval` on a previously fine-tuned model in `run_squad_albert.py`?

Thanks again, looking forward to eventually get ALBERT-xlarge model fine-tuned on SQuAD 2.0.

**Comment by pohanchi** [comment]
Hi, the performance I show didn’t modify anything and just train for that
command and wait 3 epoch then u will get the same result.

Also, to evaluate previously model, just need to modify code part which
load model to became Albert from pretrained and argument  set to the
directory name where u save your Albert model and that will work

**Comment by pohanchi** [comment]
By the way, The environment I use is one 1060 6G to produce this
performance.

**Comment by pohanchi** [comment]
The argument I use for that performance

**Comment by ahotrod** [comment]
Really good, thanks!
Your results running on a NVIDIA 1060 are impressive. 
Got a 2 x 1080Ti distributed processing script working:
```
#!/bin/bash

MODEL_PATH=/media/dn/dssd/nlp/huggingface_albert/examples/runs/albert_base_sqd1_dist_def
SQUAD_DIR=/media/dn/dssd/nlp/transformers/examples/scripts/squad1.1

python -m torch.distributed.launch --nproc_per_node=2 run_squad_albert.py \
  --train_file ${SQUAD_DIR}/train-v1.1.json \
  --predict_file ${SQUAD_DIR}/dev-v1.1.json \
  --model_type albert \
  --output_dir ${MODEL_PATH} \
  --do_train \
  --overwrite_output_dir \
  --do_lower_case \
  --warmup_steps 384 \
  --save_steps 1000 \
  --logging_steps 1000 \
  --num_train_epochs 3

CUDA_VISIBLE_DEVICES=0 python run_squad_albert.py \
  --train_file ${SQUAD_DIR}/train-v1.1.json \
  --predict_file ${SQUAD_DIR}/dev-v1.1.json \
  --model_type albert \
  --output_dir ${MODEL_PATH} \
  --do_eval \
  --do_lower_case \
  --config_name ${MODEL_PATH}
$@

***** Running training *****
Num examples = 95732
Num Epochs = 3
Instantaneous batch size per GPU = 6
Total train batch size (w. parallel, distributed & accumulation) = 96
Gradient Accumulation steps = 8
Total optimization steps = 2991

{
  "exact": 78.43897824030275,
  "f1": 86.57458158471398,
  "total": 10570,
  "HasAns_exact": 78.43897824030275,
  "HasAns_f1": 86.57458158471398,
  "HasAns_total": 10570
}
```
~33 minutes an epoch, base model occupying ~5.8 GB of 11.2 GB on each GPU.
Now to attempt ALBERT-xlarge fine-tuned on SQuAD 2.0.

**Comment by ahotrod** [comment]
The `pytorch_albert_large` 68MB & `pytorch_albert_xlarge` 225MB pytorch_state_dict downloads from google are not working.  I tried with both Linux & Win10 OSs, both hang.  The pytorch_albert (base) 45MB download works.

**Comment by pohanchi** [comment]
You need to change the pretrain model

**Comment by pohanchi** [comment]
Config pretrain also need to using

**Comment by ahotrod** [comment]
Understand I need to change the pretrained models, to either the large or xlarge pretrained models.  However, downloading the large & xlarge pretrained models from google is not working.  Only the base model downloads.

**Comment by pohanchi** [comment]
Can u show your error message ?

**Comment by ahotrod** [comment]
There is no error message.  I have albert-large running now, ~53 minutes per epoch.  Will post my results.

I was just wondering if the pre-trained models at your download link are Version 1 or Version 2 from:
[https://github.com/google-research/google-research/tree/master/albert](url)

***************New October 31, 2019 ***************
Version 2 of ALBERT models is relased. TF-Hub modules are available:
https://tfhub.dev/google/albert_base/2
https://tfhub.dev/google/albert_large/2
https://tfhub.dev/google/albert_xlarge/2
https://tfhub.dev/google/albert_xxlarge/2

Thanks so much, so far everything is working great!

**Comment by pohanchi** [comment]
Oh that’s very easy things for Version2.0, but I need to do that tomorrow
morning

**Comment by pohanchi** [comment]
Oh for the drive they are all version 1.
But version 2 I will add in the future

**Comment by ahotrod** [comment]
Thanks again for your excellent work on this.

Overnight, albert_xlarge_v1 model running ~5.5 hours per epoch, occupying ~11.1 GB of 11.2 GB on each GPU, 2 x 1080Ti:
{
  "exact": 82.22327341532639,
  "f1": 90.01828208164447,
  "total": 10570,
  "HasAns_exact": 82.22327341532639,
  "HasAns_f1": 90.01828208164447,
  "HasAns_total": 10570
}
I am running Version 2 models today.
The new `albert_config_l2.json` file has ` "num_hidden_layers": 24`,
should it be ` "num_hidden_layers": 12`?

**Comment by pohanchi** [comment]
For the google research, the config they apply on large model is 24 layer,
so for the origin version 1 on large model you need to change its layer
number to 24.

**Comment by ahotrod** [comment]
> For the google research, the config they apply on large model is 24 layer,
> so for the origin version 1 on large model you need to change its layer
> number to 24.
> …

OK, reason I asked, my script hyperparameters for V1 large model wouldn't run V2 large model, spawning GPU OOM failures.  Indeed the original paper does use 24 layer for large model.  All is well!

**Comment by pohanchi** [comment]
Your layer number also let your calculation and gpu memory increase, so 24
layer will cause OOM

**Comment by ahotrod** [comment]
`albert xlarge Version 1 SQuAD 1.1`:
{
  "exact": 82.22327341532639,
  "f1": 90.01828208164447,
  "total": 10570,
  "HasAns_exact": 82.22327341532639,
  "HasAns_f1": 90.01828208164447,
  "HasAns_total": 10570
}
`albert xlarge Version 2 SQuAD 1.1`:
{
  "exact": 0.3500473036896878,
  "f1": 7.691640650211578,
  "total": 10570,
  "HasAns_exact": 0.3500473036896878,
  "HasAns_f1": 7.691640650211578,
  "HasAns_total": 10570
}
Changed the Version 1 script to run the Version 2 model adding:
```
--model_dict_pretrain ${HFA_PATH}/pytorch_model_state_dict/pytorch_albert_xl2 \
--config_pretrain ${HFA_PATH}/config/albert_config_xl2.json \
```
Everything else in the script stayed the same.

Ref:  https://github.com/huggingface/transformers/pull/1683#issuecomment-550523247

**Comment by pohanchi** [comment]
i didn't try version2 but something will fail the em and f1 is because learning rate too big may be you can tune to 1e-5 or 2e-5
and the result will be better

**Comment by ahotrod** [comment]
Something appears to be amiss with the `albert xlarge Version 2` model:

https://github.com/google-research/google-research/issues/119#issue-520113011

No one appears to get good results, so I have been holding-off more fine-tuning Version 2 models until something is resolved.

**Comment by pohanchi** [comment]
I can double check and test about that after today midterm exam. Hope you
find solution soon.

**Comment by pohanchi** [comment]
The module I also can check. So after three hour I will do it first

**Comment by pohanchi** [comment]
@ahotrod  hey i try train albert xlarge version 2 on squad 1.1 it didn't have problem 
and EM : 85.87 F1: 92.58 still for training!!

**Comment by pohanchi** [comment]
@ahotrod  and i just try one time. what is your problem

**Comment by pohanchi** [comment]
If u need parameter and other , you can ask me I can forward it to you

**Comment by kamalkraj** [comment]
https://github.com/kamalkraj/ALBERT-TF2.0#runnig-squad-v20

**Comment by ahotrod** [comment]
> If u need parameter and other , you can ask me I can forward it to you
> 
> On Sun, Nov 10, 2019 at 15:02 aapp1420@gmail.com.tw <aapp1420@gmail.com>
> wrote:
> …

I'll try another fine-tuning run this week.  It would be helpful to see your shell script for your albert xlarge version 2 on squad 1.1.  Are you running this on the 6GB 1060 as previously mentioned??

**Comment by pohanchi** [comment]
I run on 1 gpu 1080Ti batch_size 2  accumulate 24 eat 7G memory

**Comment by ahotrod** [comment]
> https://github.com/kamalkraj/ALBERT-TF2.0#runnig-squad-v20

@kamalkraj I looked at your work, excellent, thanks.

As the Huggingface Transformer paper describes on page 2, 5th paragraph, I'm a "low-resource user", with only a 2x 1080 Ti(s) system.  For me to run the larger ALBERT-TF2.0 models, I would need gradient accumulation steps, and I didn't see that option in your `run_squad.py` code.

**Comment by pohanchi** [comment]
On argparser, you will find the parameter with accumulated_step or other
parameter called accumulated....

**Comment by kamalkraj** [comment]
> > https://github.com/kamalkraj/ALBERT-TF2.0#runnig-squad-v20
> 
> I looked at your work, excellent, thanks.
> 
> As the Huggingface Transformer paper describes on page 2, 5th paragraph, I'm a "low-resource user", with only a 2x 1080 Ti(s) system. For me to run the larger ALBERT-TF2.0 models, I would need gradient accumulation steps, and I didn't see that option in your `run_squad.py` code.

I will try adding that option (gradient accumulation). Thanks for the feedback @ahotrod

**Comment by pohanchi** [comment]
python3 run_squad_albert.py --train_file ../data/train-v1.1.json
--predict_file ../data/dev-v1.1.json --model_type albert --output_dir pohan
--config_pretrain config/albert_config_xl2.json --model_dict_pretrain
./pytorch_model_state_dict/pytorch_albert_xl2 --do_train --do_eval
--do_lower_case --evaluate_during_training --per_gpu_train_batch_size 2
--per_gpu_eval_batch_size 2 --learning_rate 1e-5
--gradient_accumulation_steps 24 --weight_decay 0.1 --adam_epsilon 1e-12
--num_train_epochs 3 --warmup_steps 1000 --logging_steps 1000 --save_steps
1000 --fp16

Kamal Raj <notifications@github.com> 於 2019年11月11日 週一 上午1:27寫道：

**Comment by ahotrod** [comment]
> python3 run_squad_albert.py --train_file ../data/train-v1.1.json
> --predict_file ../data/dev-v1.1.json --model_type albert --output_dir pohan
> --config_pretrain config/albert_config_xl2.json --model_dict_pretrain
> ./pytorch_model_state_dict/pytorch_albert_xl2 --do_train --do_eval
> --do_lower_case --evaluate_during_training --per_gpu_train_batch_size 2
> --per_gpu_eval_batch_size 2 --learning_rate 1e-5
> --gradient_accumulation_steps 24 --weight_decay 0.1 --adam_epsilon 1e-12
> --num_train_epochs 3 --warmup_steps 1000 --logging_steps 1000 --save_steps
> 1000 --fp16
> 
> Kamal Raj <notifications@github.com> 於 2019年11月11日 週一 上午1:27寫道：
> …

@pohanchi Thanks so much, most helpful.  I'll give it a try this week.

**Comment by ahotrod** [comment]
@pohanchi Should have time to make some test runs over the next few days.

I would like to run `pytorch_albert_xxlarge` (Version 1) but there's no entry in `pytorch_model_state_dict`.  Is that something I can generate?  Will also probably need `albert_config_xxlarge.json` in `config`.

**Comment by pohanchi** [comment]
Oh I think xxlarge Model is not efficiency for me(low resource),so i didn’t want to generate! But if u need, I can generate it for a few day later

**Comment by pohanchi** [comment]
Another method to do this, just go to version 1 tf-hub web page for albert xxlarge and download the tf-hub module and using my convert_tf_xxx.py to generate model_state_dict . Config you can just copy config on tf-hub module webpage and generate a Json file

**Comment by pohanchi** [comment]
@ahotord I will do it this week but not in hurry. So maybe it will in Friday or Saturday. By the way, I add Lamb optimizer and it perform normal. But it can use !

**Comment by ahotrod** [comment]
> Another method to do this, just go to version 1 tf-hub web page for albert xxlarge and download the tf-hub module and using my convert_tf_xxx.py to generate model_state_dict . Config you can just copy config on tf-hub module webpage and generate a Json file

@pohanchi
Perfect, I'll give this a try.

Per the original paper's github, albert_xxlarge V1 outperforms V2 new release.
I'm currently running albert_xlarge with bs=2 per gpu, so I figure I can run xxlarge with bs=1 and 2x grad_accumulation (24).  Looking for every bit of increased "accuracy" for fine-tuned SQuAD 2.0.

Thanks again.

**Comment by ahotrod** [comment]
`albert xlarge Version 1 SQuAD 2.0`:
{
  "exact": 81.58847805946264,
  "f1": 84.95025543030565,
  "total": 11873,
  "HasAns_exact": 78.12078272604589,
  "HasAns_f1": 84.85397819231112,
  "HasAns_total": 5928,
  "NoAns_exact": 85.04625735912532,
  "NoAns_f1": 85.04625735912532,
  "NoAns_total": 5945,
  "best_exact": 81.93379937673714,
  "best_exact_thresh": -1.7692835330963135,
  "best_f1": 85.16188049847581,
  "best_f1_thresh": -0.575573205947876
}
`albert xlarge Version 2 SQuAD 2.0`, significant improvement:
{
  "exact": 84.51107554956624,
  "f1": 87.57064747649119,
  "total": 11873,
  "HasAns_exact": 80.70175438596492,
  "HasAns_f1": 86.82967231585373,
  "HasAns_total": 5928,
  "NoAns_exact": 88.30950378469302,
  "NoAns_f1": 88.30950378469302,
  "NoAns_total": 5945,
  "best_exact": 84.61214520340268,
  "best_exact_thresh": -1.5449519157409668,
  "best_f1": 87.62058069832646,
  "best_f1_thresh": -0.16038274765014648
}
Unable to fit `albert xxlarge Version 1` on single GPU with bs=1.  Will eventually use cloud V100s for fine-tuning that model on SQuAD 2.0.

**Comment by pohanchi** [comment]
So they all can train, right. They are no problems.

**Comment by ahotrod** [comment]
> So they all can train, right. They are no problems.

@pohanchi `albert xlarge` Versions 1 & 2 train & evaluate on SQuAD 1.1 & 2.0, thanks!
`albert xlarge Version 2 SQuAD 2.0` will be my model of choice for now.
